backend/controller/vm.js

```diff
@@ -0,0 +1,130 @@
+
+    if (serviceId == null || userAddr == null || duration == null) {
+        throw new BadRequestError("Missing required fields");
+    }
+
+    if (serviceId !== virtualMachineServiceId) {
+        throw new BadRequestError("Invalid serviceId");
+    }
+
+    // Step 1: Generate SSH Key Pair
+    const keyGenConfig = {
+        location: path.join(__dirname, `../keyPairs/${serviceId + userAddr}`),
+        type: "rsa",
+        read: true,
+        force: true,
+        destroy: true,
+        comment: "joe@foobar.com",
+        // password: "keypassword",
+        size: "2048",
+        format: "PEM",
+    };
+    sshKeygen(keyGenConfig, async (err, keyPair) => {
+        if (err) {
+            console.log(err);
+            return res
+                .status(500)
+                .json({ msg: "Failed to generate SSH key pair." });
+        }
+
+        // Step 2: Use Azure SDK to Create VM (refer to previous response)
+        await createAzureVm(serviceId, userAddr, duration, keyPair);
+        // Step 3: Convert Private Key to .pem Format
+        console.log(keyPair);
+        const privateKeyPEM = sshpk.parsePrivateKey(keyPair.key, "pem");
+        const privateKeyPEMString = privateKeyPEM.toString("pem");
+
+        // Step 4: Send Response with Private Key .pem
+        res.setHeader("Content-Type", "application/x-pem-file");
+        res.setHeader(
+            "Content-Disposition",
+            "attachment; filename=private_key.pem"
+        );
+        res.send(privateKeyPEMString);
+    });
+};
+
+const getVmDomain = async (req, res) => {
+    // Azure authentication in environment variables for DefaultAzureCredential
+    const tenantId = process.env.AZURE_TENANT_ID;
+    const clientId = process.env.AZURE_CLIENT_ID;
+    const secret = process.env.AZURE_CLIENT_SECRET;
+    const subscriptionId = process.env.AZURE_SUBSCRIPTION_ID;
+
+    const { serviceId, userAddr } = req.body;
+
+    let credentials = null;
+    credentials = new ClientSecretCredential(tenantId, clientId, secret);
+
+    const networkClient = new NetworkManagementClient(
+        credentials,
+        subscriptionId
+    );
+    const computeClient = new ComputeManagementClient(
+        credentials,
+        subscriptionId
+    );
+    try {
+        const { dnsSettings } = await networkClient.publicIPAddresses.get(
+            `${serviceId + userAddr}-testrg`,
+            `${serviceId + userAddr}-pip`
+        );
+        const { osProfile } = await computeClient.virtualMachines.get(
+            `${serviceId + userAddr}-testrg`,
+            `${serviceId + userAddr}vm`
+        );
+
+        res.json({ dns: dnsSettings.fqdn, username: osProfile.adminUsername });
+    } catch (error) {
+        // if (error.code === "ResourceGroupNotFound") {
+        //     res.status(404).json({ msg: "Resource not found" });
+        // } else {
+        throw new BadRequestError(error.message);
+        // }
+    }
+};
+
+const stopVm = async (req, res) => {
```

Comment:
this should also be authenticated by `signature`

Reply:
so whe its time i would create a middleware that would handle the auth with signature

Reply:
alright.